2023/incoming/the-magical-test2-suite.pod

```diff
@@ -0,0 +1,159 @@
+        ingredients => {
+            eggs => 5,
+            flour => 3,
+            'dried fruit' => ['cherries', 'apricots', 'dates'],
+        },
+    }, 'got proper dessert!';
+
+What about that time of the year when you get a data structure or object
+and care only about a few keys, items or attributes? Ooh, Santa’s got a
+gift for you, too! B<< Use like() with the nested structure to ignore any keys/positions you haven’t defined >> in a true non-strict (partial) match.
+It even lets you mix and match between exact values (by passing a string
+or a number) and values that match a regular expression (by providing
+the regexp).
+
+    like $recipe, {
+        name => qr(cake)i,  # must contain ‘cake’ (case insensitive)
+        ingredients => {
+            eggs => qr(\d+),   # any number
+            flour => 5,            # exactly 5
+            dried fruit’ => [qr/cherr(y|ies)/, 'apricots', qr/dates?/],  # mix and match!
+        }
+    }, 'partial match in nested variables, mixing is() and like() at any level';
+```
+
+Think the presents are over? Think again! For even more complex validations you can check your variable against a builder (and there are L<< many builders available for hashes, arrays, objects, etc|https://metacpan.org/pod/Test2::Tools::Compare#VALUE-SPECIFICATIONS >>). For example, let’s say I wanted
+to check whether C<$recipe> has a name and ingredients, and if one of the
+dried fruits is raisins. Also, just to make it a little harder, let’s
+make sure it has no key called ‘microwave’. To do all that, we just write
+a very simple definition of our partial hash containing only the bits we
+care about:
+
+    # import everything we use in this test
+    # (the :DEFAULT label is to ensure all regular symbols are also imported)
+    use Test2::V0 qw( :DEFAULT hash field bag item etc L DNE );
+
+    is $recipe, hash {
+      field name => L;  # the value of the 'name' key is defined and has a L()ength.
+      field ingredients => hash {
+        # 'bag' is an 'array' that doesn't care about element order.
+        field 'dried fruit' => bag { item 'raisins'; etc; };
+        etc;
+      };
+      field microwave => DNE; # the 'microwave' key Does Not Exist.
+      etc;    # ignore other keys. Use 'end' to fail the test if other keys exist.
+    }, 'partial match from a generated definition!';
+
+If that wasn’t impressive enough, here are some extra nice ways to make your tests more thorough, robust and clear without having to load external modules or fiddle with the symbol table:
+
+=head2 Test if loading a module imports (or doesn’t import) a function or a variable:
+
+    use Some::Module;
+    imported_ok 'mysub', '$myvar', '@myothervar';
+    not_imported_ok 'othersub', '$othervar';
+```
+
+=head2 Test if something warns or dies / raises an exception
+
+    like dies { … }, qr/some error/, ‘got expected exception from block’;
+    ok lives { … }, ‘code lived!’, “oh, noes! Died with error ‘$@’”;
```

Comment:
Verify this one. Most of the time the Test2 tools localize $@ so that they cannot modify it. I am not sure if this tool preserves $@ so that it can be used in the debuging like that.

Reply:
I did, and looks like it works:

```perl
use Test2::V0;
sub foo { die 42 }
ok lives { foo() }, 'code lived!', "oh noes! Died with error '$@'";
done_testing
```

```
$ yath test.pl 

** Defaulting to the 'test' command **

[  FAIL  ]  job  1  + code lived!
(  DIAG  )  job  1    Failed test 'code lived!'
(  DIAG  )  job  1    at test2.pl line 5.
(  DIAG  )  job  1    ----- START -----
oh noes! Died with error '42 at test2.pl line 3.
'
(  DIAG  )  job  1    ------ END ------
(  DIAG  )  job  1    Seeded srand with seed '20231210' from local date.
( FAILED )  job  1    test2.pl
< REASON >  job  1    Test script returned error (Err: 1)
< REASON >  job  1    Assertion failures were encountered (Count: 1)

The following jobs failed:
+--------------------------------------+-----------+
| Job ID                               | Test File |
+--------------------------------------+-----------+
| 422AE85C-9785-11EE-AECC-7AD9F454D40C | test2.pl  |
+--------------------------------------+-----------+

                                Yath Result Summary
-----------------------------------------------------------------------------------
     Fail Count: 1
     File Count: 1
Assertion Count: 1
      Wall Time: 0.49 seconds
       CPU Time: 0.75 seconds (usr: 0.22s | sys: 0.04s | cusr: 0.42s | csys: 0.07s)
      CPU Usage: 153%
    -->  Result: FAILED  <--
```
As you can see, `$@` shows up properly on the output. But I know it's quite the beast, so I rather have your eyes on it than mine, maybe I missed something.